// mora-common/widget/SideEffect.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var withSideEffect = require("react-side-effect");
var React = require("react");
var SideEffect = /** @class */ (function (_super) {
    tslib_1.__extends(SideEffect, _super);
    function SideEffect() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    SideEffect.prototype.render = function () {
        var children = this.props.children;
        return children ? React.Children.only(children) : null;
    };
    return SideEffect;
}(React.PureComponent));
exports.SideEffect = SideEffect;
var BodyClassName = /** @class */ (function (_super) {
    tslib_1.__extends(BodyClassName, _super);
    function BodyClassName() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    BodyClassName = tslib_1.__decorate([
        withSideEffect(makeReducePropsToState('className'), function (className) {
            var nextClassname = className || '';
            if (nextClassname !== document.body.className)
                document.body.className = nextClassname;
        })
    ], BodyClassName);
    return BodyClassName;
}(SideEffect));
exports.BodyClassName = BodyClassName;
// react-document-title v2.0.3
var DocumentTitle = /** @class */ (function (_super) {
    tslib_1.__extends(DocumentTitle, _super);
    function DocumentTitle() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    DocumentTitle = tslib_1.__decorate([
        withSideEffect(makeReducePropsToState('title'), function (title) {
            var nextTitle = title || '';
            if (nextTitle !== document.title)
                document.title = nextTitle;
        })
    ], DocumentTitle);
    return DocumentTitle;
}(SideEffect));
exports.DocumentTitle = DocumentTitle;
function makeReducePropsToState(key) {
    return function (propsList) {
        var innermostProps = propsList[propsList.length - 1];
        if (innermostProps) {
            return innermostProps[key];
        }
    };
}
