// @minapp/common/dist/lib/CacheableFile.js
"use strict";
/******************************************************************
 MIT License http://www.opensource.org/licenses/mit-license.php
 Author Mora <qiuzhongleiabc@126.com> (https://github.com/qiu8310)
*******************************************************************/
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var fs_1 = require("./fs");
/**
 * 支持缓存的文件
 *
 * 避免频繁去读取文件系统
 *
 * @export
 * @class File
 */
var CacheableFile = /** @class */ (function () {
    function CacheableFile(filepath) {
        this.filepath = filepath;
    }
    /**
     * 获取文件内容(如果文件没有更新过，优先使用缓存)
     *
     * @returns Promise<Buffer>
     * @memberof CacheableFile
     */
    CacheableFile.prototype.getContent = function () {
        return __awaiter(this, void 0, void 0, function () {
            var stats, ctime, mtime, _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, fs_1.stat(this.filepath)];
                    case 1:
                        stats = _b.sent();
                        if (!stats.isFile())
                            throw new Error(this.filepath + " is not a file");
                        ctime = stats.ctime.getTime();
                        mtime = stats.mtime.getTime();
                        if (!(!this.ctime || !this.mtime || !this.cachedContent || this.ctime !== ctime || this.mtime !== mtime)) return [3 /*break*/, 3];
                        this.ctime = ctime;
                        this.mtime = mtime;
                        _a = this;
                        return [4 /*yield*/, fs_1.readFile(this.filepath)];
                    case 2:
                        _a.cachedContent = _b.sent();
                        this.cached = false;
                        return [3 /*break*/, 4];
                    case 3:
                        this.cached = true;
                        _b.label = 4;
                    case 4: return [2 /*return*/, this.cachedContent];
                }
            });
        });
    };
    return CacheableFile;
}());
exports.CacheableFile = CacheableFile;
