// mora-common/dom/getLocation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var url_1 = require("../util/url");
var cache = {};
function getLocation(options) {
    var _a = window.location, pathname = _a.pathname, search = _a.search, hash = _a.hash;
    var l = cache.l, p = cache.p, s = cache.s, h = cache.h;
    if (l && p === pathname && s === search && h === hash)
        return l;
    cache = { l: _getLocation(options), p: pathname, s: search, h: hash };
    return cache.l;
}
exports.getLocation = getLocation;
function _getLocation(options) {
    var _a = window.location, pathname = _a.pathname, search = _a.search, hash = _a.hash;
    var href;
    var query;
    if (options.hash) {
        href = location.hash.substr((options.separater || '#').length);
        pathname = href.split(/[?#]/).shift() || '';
        search = href.substr(pathname.length).split(/#/).shift() || '';
        hash = href.indexOf('#') >= 0 ? href.substr(href.indexOf('#')) : '';
    }
    query = url_1.parseSearch(search);
    search = url_1.buildSearch(query);
    return {
        url: pathname + search + hash,
        search: search,
        pathname: pathname,
        nativeQuery: options.hash ? url_1.parseSearch(location.search) : query,
        query: query,
        hash: hash
    };
}
