// mora-common/dom/KeyCode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
function expand(codes) {
    var i;
    // lower case chars
    for (i = 97; i < 123; i++)
        codes[String.fromCharCode(i)] = i - 32;
    // numbers
    for (i = 48; i < 58; i++)
        codes[i - 48] = i;
    return codes;
}
// 不要出现重复的 value
exports.FunctionKeyCode = {
    backspace: 8,
    tab: 9,
    enter: 13,
    shift: 16,
    ctrl: 17,
    alt: 18,
    esc: 27,
    space: 32,
    pgup: 33,
    pgdn: 34,
    home: 36,
    left: 37,
    up: 38,
    right: 39,
    down: 40,
    insert: 45,
    delete: 46,
    // windows: 91,
    // command: 91,
    // leftCommand: 91,
    lmeta: 91,
    // rightCommand: 93,
    rmeta: 93,
    f1: 112, f2: 113, f3: 114, f4: 115, f5: 116, f6: 117, f7: 118, f8: 119, f9: 120, f10: 121, f11: 122
};
exports.PrintKeyCode = expand({
    ';': 186,
    '=': 187,
    ',': 188,
    '-': 189,
    '.': 190,
    '/': 191,
    '`': 192,
    '[': 219,
    '\\': 220,
    ']': 221,
    '"': 222
});
exports.KeyCode = tslib_1.__assign({}, exports.FunctionKeyCode, exports.PrintKeyCode);
