// mora-common/helper/index-loader/inc/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/** 匹配带 export 的所有行 */
exports.exportLineRegexp = /^export\s+.*?$/gm;
/**
 * 匹配 export = xxx
 */
exports.lineExportAllRegExp = /^export\s+=\s+(\w+)/;
/**
 * 匹配 export { a, b as c } from 'xxx';
 *
 * $1 是 " a, b as c "
 *
 * $2 是 "xxx"
 */
exports.lineExportLocalsFromRegexp = /^export\s+\{([^}]*?)\}\s+from\s+['"](.*?)['"]/;
/**
 * 匹配 export * from 'xxx'
 *
 * $1 是 "*"
 *
 * $2 是 'xxx'
 */
exports.lineExportAllFromRegExp = /^export\s+(\*)\s+from\s+['"](.*?)['"]/;
/** 匹配 export default */
exports.lineExportDefaultRegexp = /^export\s+(default)\s+/;
/** 匹配的变量的声明，如 export interface xxx,  export class yyy */
exports.lineExportVariableRegexp = /^export\s+(?:declare\s+)?(?:abstract\s+)?(?:async\s+)?(?:class|type|function|interface|const|let)\s+(\w+)/;
/** 匹配的变量的声明，如 export {a, b as c}，注意和 lineExportLocalsFromRegexp 区分，这里没有 from */
exports.lineExportLocalsRegexp = /^export\s+\{([^}]*?)\}/;
/** 匹配带 import 的所有行 */
exports.importLineRegExp = /^import\s+.*?$/gm;
/** 匹配 import * as xxx from './xxx' */
exports.lineImportAllRegExp = /^import\s+\*\s+as\s+(\w+)\s+from\s+['"](.*?)['"]/;
/**
 * 匹配
 *
 * export namespace xxx {
 *  // ....
 * }
 *
 * 会将 namespace 中的所有 export 清空，所以要保证当前有个同名的 export 存在
 */
exports.exportNamespaceRegExp = /^export\s+(?:declare\s+)?namespace[\s\S]*?\n\}/gm;
/**
 * 匹配 import xxx from './xxx'
 *
 * 还需要去检查 from 的文件内容是否有 export default，因为 tsconfig 可以开启 allowSyntheticDefaultImports
 */
exports.lineImportDefaultRegExp = /^import\s+(\w+)\s+from\s+['"](.*?)['"]/;
/** 匹配 import {a, b as c} from './xxx' */
exports.lineImportLocalsRegExp = /^import\s+\{([^}]*?)\}\s+from\s+['"](.*?)['"]/;
/** 匹配 import './index.scss' 这种只引用的形式 */
exports.lineImportOnlyRegExp = /^import\s+['"](.*?)['"]/;
/** 匹配 import React, { Component } from 'react'， 需要 tsconfig 开启 allowSyntheticDefaultImports */
exports.lineImportAllAndLocalsRegExp1 = /^import\s+(\w+)\s*\,\s*\{[^}]*?\}\s+from\s+['"](.*?)['"]/;
/** 匹配 import { Component }, React from 'react'， 需要 tsconfig 开启 allowSyntheticDefaultImports */
exports.lineImportAllAndLocalsRegExp2 = /^import\s+\{[^}]*?\}\s*,\s*(\w+)\s+from\s+['"](.*?)['"]/;
/**
 * 拆分 locals 里的 "a, b as c" 这种字符串
 *
 * 注意 split 最好前先 trim 字符串，并且确认字符串有值
 */
exports.splitRegexp = /\s*,\s*/;
/** */
exports.asRegexp = /^(\w+)\s+as\s+(\w+)$/;
/** 放在文件路径后面的分隔符，后面接文件中的变量 */
exports.KEY_SEPARATOR = '::';
/** 放在 KEY_SEPARATOR 后面，表示导出此文件的所有变量 */
exports.KEY_ALL = '*';
exports.KEY_DEFAULT = 'default';
