// mora-common/helper/index-loader/inc/fn.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var findup = require("mora-scripts/libs/fs/findup");
var exists = require("mora-scripts/libs/fs/exists");
var _warn = require("mora-scripts/libs/sys/warn");
var _info = require("mora-scripts/libs/sys/info");
var _error = require("mora-scripts/libs/sys/error");
var path = require("path");
var envCache = [];
/**
 * 根据项目中的某一个文件，获取当前项目的一些基本信息
 */
function env(srcFile) {
    for (var _i = 0, envCache_1 = envCache; _i < envCache_1.length; _i++) {
        var cache = envCache_1[_i];
        if (srcFile.indexOf(cache.rootDir) === 0)
            return cache;
    }
    var pkgFile = findup.pkg(path.dirname(srcFile));
    var result = {
        // pkg: require(pkgFile),
        rootDir: path.dirname(pkgFile)
    };
    envCache.push(result);
    return result;
}
exports.env = env;
/**
 * 祛除单行中的注释
 */
function stripInlineComment(line) {
    return line.replace(/\/\*.*?\*\//g, '');
}
exports.stripInlineComment = stripInlineComment;
function isFileExists(file) {
    return exists(file);
}
exports.isFileExists = isFileExists;
function warn() {
    var args = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        args[_i] = arguments[_i];
    }
    _warn.apply(void 0, ['[index-loader]'].concat(args));
}
exports.warn = warn;
function info() {
    var args = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        args[_i] = arguments[_i];
    }
    _info.apply(void 0, args);
}
exports.info = info;
function error() {
    var args = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        args[_i] = arguments[_i];
    }
    _error.apply(void 0, ['[index-loader]'].concat(args));
}
exports.error = error;
